src/engine/renderer/MaterialCommon.h

```diff
@@ -0,0 +1,242 @@
+{
+	return static_cast<size_t>( GetLightMapNum( obj ) ) < tr.deluxemaps.size();
+}
+
+template<typename Obj> image_t* GetDeluxeMap( Obj* obj )
+{
+	if ( HasDeluxeMap( obj ) )
+	{
+		return tr.deluxemaps[ GetLightMapNum( obj ) ];
+	}
+
+	return tr.blackImage;
+}
+
+inline shader_t* GetSurfaceShader( shaderCommands_t* tess )
+{
+	return tess->surfaceShader;
+}
+
+inline shader_t* GetSurfaceShader( drawSurf_t* drawSurf )
+{
+	return drawSurf->shader;
+}
+
+template<typename Obj> static bool hasExplicitelyDisabledLightMap( Obj* obj )
+{
+	return GetSurfaceShader( obj )->surfaceFlags & SURF_NOLIGHTMAP;
+}
+
+inline shaderStage_t* GetSurfaceLastStage( shaderCommands_t* tess )
+{
+	return tess->surfaceLastStage;
+}
+
+inline shaderStage_t* GetSurfaceLastStage( drawSurf_t* drawSurf )
+{
+	return drawSurf->shader->lastStage;
+}
+
+inline shaderStage_t* GetSurfaceStages( shaderCommands_t* tess )
+{
+	return tess->surfaceStages;
+}
+
+inline shaderStage_t* GetSurfaceStages( drawSurf_t* drawSurf )
+{
+	return drawSurf->shader->stages;
+}
+
+template<typename Obj> bool isExplicitelyVertexLitSurface( Obj* obj )
+{
+	shaderStage_t* lastStage = GetSurfaceLastStage( obj );
+	shaderStage_t* stages = GetSurfaceStages( obj );
+	return lastStage != stages && stages[0].rgbGen == colorGen_t::CGEN_VERTEX;
+}
+
+template<typename Obj> void SetLightDeluxeMode( Obj* obj,
+	stageType_t stageType,
+	lightMode_t& lightMode, deluxeMode_t& deluxeMode )
+{
+	lightMode = lightMode_t::FULLBRIGHT;
+	deluxeMode = deluxeMode_t::NONE;
+
+	if ( hasExplicitelyDisabledLightMap( obj ) && !isExplicitelyVertexLitSurface( obj ) )
+	{
+		// Use fullbright on “surfaceparm nolightmap” materials.
+	}
+	else if ( stageType == stageType_t::ST_COLLAPSE_COLORMAP )
+	{
+		/* Use fullbright for collapsed stages without lightmaps,
+		for example:
+
+		  {
+		    map textures/texture_d
+		    heightMap textures/texture_h
+		  }
+
+		This is doable for some complex multi-stage materials. */
+	}
+	else if ( obj->bspSurface )
+	{
+		lightMode = tr.worldLight;
+		deluxeMode = tr.worldDeluxe;
+
+		if ( lightMode == lightMode_t::MAP )
+		{
+			if ( !HasLightMap( obj ) )
+			{
+				lightMode = lightMode_t::VERTEX;
+				deluxeMode = deluxeMode_t::NONE;
+			}
+		}
+	}
+	else
+	{
+		lightMode = tr.modelLight;
+		deluxeMode = tr.modelDeluxe;
+	}
+}
+
+template<typename Obj> image_t* SetLightMap( Obj* obj, lightMode_t lightMode )
+{
+	switch ( lightMode ) {
+		case lightMode_t::VERTEX:
+			break;
+
+		case lightMode_t::GRID:
+			return tr.lightGrid1Image;
+			break;
+
+		case lightMode_t::MAP:
+			return GetLightMap( obj );
+			break;
+
+		default:
+			break;
+	}
+
+	return tr.whiteImage;
+}
+
+template<typename Obj> image_t* SetDeluxeMap( Obj* obj, deluxeMode_t deluxeMode )
+{
+	switch ( deluxeMode ) {
+		case deluxeMode_t::MAP:
+			return GetDeluxeMap( obj );
+			break;
+
+		case deluxeMode_t::GRID:
+			return tr.lightGrid2Image;
+			break;
+
+		default:
+			break;
+	}
+
+	return tr.blackImage;
```

Comment:
This defaulted to `tr.whiteImage` in the original code, and might not work correctly otherwise.

Reply:
It seems the difference between 0 and 1 would be that lightDir would be either (1.0, 1.0, 1.0) or (-1.0, -1.0, -1.0), but I haven't checked further if it makes any difference.

Reply:
If this is supposed to be a direction then there is this:
```image_t    *flatImage; // use this as default normalmap```

Is this even used though? I would imagine if we have `deluxeMode_t::NONE` then the deluxe-using shaders should just be disabled.

Reply:
The normal map flat image is probably no better at is is `{0.5, 0.5, 1.0}` and here what we may look for may be `{0.5, 0.5, 0.5}`. Anyway, the image is likely not used at all when the feature is not available, it is likely set to avoid handling a null pointer.